### DISCUSSION TITLE
Behavior of `fgetc` in `tinystdio` version regarding error flags

### DISCUSSION THREAD

**Discussion by M-Moawad** [picolibc/picolibc#717] [Q&A] [answered]
Hello everyone,

I'd like to initiate a discussion regarding the behavior of the `fgetc` function in the tinystdio version of the Picolibc library, specifically concerning the handling of error flags on stream flags.

Here's a code snippet to demonstrate this test case:

```c
#include <stdio.h>

int main() {
    FILE *file;
    int ch;
    int flags_before, flags_after;

    // Open file in writing mode
    file = fopen("test.txt", "w");
    if (file == NULL) {
        perror("Error opening file");
        return 1;
    }

    // Get file descriptor flags before reading
    flags_before = file->flags;

    // Attempt to read from the file
    ch = fgetc(file);
    if (ch == EOF) {
        printf("Reached end of file or error occurred while reading.\n");
    }

    // Get file descriptor flags after reading
    flags_after = file->flags;

    // Print flags before and after
    printf("File descriptor flags before fgetc: %d\n", flags_before);
    printf("File descriptor flags after fgetc: %d\n", flags_after);

    if ((flags_after & __SERR) == 1) {
        printf("Test case passed.\n");
    } else {
        printf("Test case failed.\n");
    }

    // Close the file
    fclose(file);

    return 0;
}
```

According to the previous test case:

  1. **Expected Behavior:** In standard C libraries, attempting to read from a file is not opened in read mode, results in setting the error indicator on the stream, and returns EOF. **ISO_IEC_9899_1999 section 7.19.7.1**
  2. **Observation in tinystdio:** In my test with the tinystdio version of Picolibc, I found that the error flag was not raised when attempting to read from a file opened in write mode.
  Checking the code of `fgetc` here: https://github.com/picolibc/picolibc/blob/main/newlib/libc/tinystdio/fgetc.c
  ```c
  int
fgetc(FILE *stream)
{
	int rv;
	__ungetc_t unget;

	if ((stream->flags & __SRD) == 0)     // <-- in this check, if the steam->flags don't have the __SRD set, `fgetc` returns EOF directly without raising an error flag 
		return EOF;
```

Looking forward to your contributions!

Best regards,
Mohamed Moawad

**Comment by keith-packard** [comment, marked as answer]
I've reviewed the C17 and POSIX specifications for fgetc and agree that the error indicator should be set in this case. Do you want to propose a patch along with a test case? The above code would need to be changed to use `ferror()` instead of directly probing the contents of the file, but otherwise it looks like it would make a good test case.

**Reply by M-Moawad** [reply]
Thank you for your quick response, I've updated the test case to use `ferror()` instead of directly probing the contents of the file. You can find the updated test case below:
```c
#include <stdio.h>

int main() {
    FILE *file;
    int ch;
    int error_before, error_after;

    // Open file in writing mode
    file = fopen("test.txt", "w");
    if (file == NULL) {
        perror("Error opening file");
        return 1;
    }

    // Get error indicator before reading
    error_before = ferror(file);

    // Attempt to read from the file
    ch = fgetc(file);
    if (ch == EOF) {
        if (ferror(file)) {
            printf("An error occurred while reading.\n");
        } else {
            printf("Reached end of file.\n");
        }
    }

    // Get error indicator after reading
    error_after = ferror(file);

    // Print error indicators before and after
    printf("Error indicator before fgetc: %d\n", error_before);
    printf("Error indicator after fgetc: %d\n", error_after);

    if (error_after != 0) {
        printf("Test case passed.\n");
    } else {
        printf("Test case failed.\n");
    }

    // Close the file
    fclose(file);

    return 0;
}
```

If you have any assumptions or corner cases not clear to me and would like to apply a patch for this issue, please feel free to add it
